src/game/Utils/Cinematics.cc

```diff
@@ -7,200 +7,313 @@
 };
 
 
-// SMKFLIC uiFlags
+// SMKFLIC flags
 #define SMK_FLIC_OPEN      0x00000001 // Flic is open
 #define SMK_FLIC_PLAYING   0x00000002 // Flic is playing
 #define SMK_FLIC_AUTOCLOSE 0x00000008 // Close when done
 
-static SMKFLIC SmkList[4];
-static UINT32  guiSmackPixelFormat = SMACKBUFFER565;
-
 
-BOOLEAN SmkPollFlics(void)
-{
-	BOOLEAN fFlicStatus = FALSE;
-	FOR_EACH(SMKFLIC, i, SmkList)
-	{
-		if (!(i->uiFlags & SMK_FLIC_PLAYING)) continue;
-		fFlicStatus = TRUE;
+static SMKFLIC gSmkList[4];
 
-		Smack* const smkobj = i->SmackerObject;
 
+static SMKFLIC* SmkOpenFlic(const char* filename);
+static SMKFLIC* SmkGetFreeFlic(void);
+static void SmkSkipFrames(SMKFLIC* sf);
+static void SmkBlitVideoFrame(SMKFLIC* const sf, SGPVSurface* surface);
 
-		if (SmackWait(smkobj)) continue;
 
-		//if (SmackSkipFrames(smkobj)) continue;
+BOOLEAN SmkPollFlics(void)
+{
+	BOOLEAN is_playing = FALSE;
+	FOR_EACH(SMKFLIC, sf, gSmkList)
+	{
+		if (!(sf->flags & SMK_FLIC_PLAYING)) continue;
 
-		{
-			SGPVSurface::Lock l(FRAME_BUFFER);
-			SmackToBuffer(smkobj, i->uiLeft, i->uiTop, l.Pitch(), smkobj->Height, smkobj->Width, l.Buffer<UINT16>(), guiSmackPixelFormat);
-			SmackDoFrame(smkobj);
-		}
+		SmkSkipFrames(sf);
 
-		// Check to see if the flic is done the last frame
-		//printf ("smk->FrameNum %u\n", smk->FrameNum);
-		// if (smk->FrameNum == smk->Frames - 1)
-		if (i->SmackerStatus == SMK_LAST )
+		if (sf->status == SMK_DONE || sf->status == SMK_ERROR)
 		{
-			if (i->uiFlags & SMK_FLIC_AUTOCLOSE) SmkCloseFlic(i);
+			if (sf->flags & SMK_FLIC_AUTOCLOSE)
+			{
+				SmkCloseFlic(sf);
+			}
+			else
+			{
+				sf->status &= ~SMK_FLIC_PLAYING;
+			}
 		}
 		else
 		{
-			i->SmackerStatus = SmackNextFrame(smkobj);
+			is_playing = TRUE;
+			SmkBlitVideoFrame(sf, FRAME_BUFFER);
 		}
 	}
 
-	return fFlicStatus;
+	return is_playing;
 }
 
 
 void SmkInitialize(void)
 {
-	// Wipe the flic list clean
-	memset(SmkList, 0, sizeof(SmkList));
-
-	// Use MMX acceleration, if available
-	SmackUseMMX(1);
+	// assume the list contains garbage
+	FOR_EACH(SMKFLIC, sf, gSmkList)
+	{
+		sf->file_in_memory = nullptr;
+		sf->smacker = nullptr;
+		FOR_EACH(UINT32, sound, sf->sounds)
+		{
+			*sound = NO_SAMPLE;
+		}
+		sf->flags = 0;
+	}
 }
 
 
 void SmkShutdown(void)
 {
-	// Close and deallocate any open flics
-	FOR_EACH(SMKFLIC, i, SmkList)
+	FOR_EACH(SMKFLIC, sf, gSmkList)
 	{
-		if (i->uiFlags & SMK_FLIC_OPEN) SmkCloseFlic(i);
+		SmkCloseFlic(sf);
 	}
 }
 
 
-static SMKFLIC* SmkOpenFlic(const char* filename);
-
-
 SMKFLIC* SmkPlayFlic(const char* const filename, const UINT32 left, const UINT32 top, const BOOLEAN auto_close)
 {
+	char status;
 	SMKFLIC* const sf = SmkOpenFlic(filename);
-	if (sf == NULL) return NULL;
+	if (sf == nullptr) return nullptr;
 
 	// Set the blitting position on the screen
-	sf->uiLeft = left;
-	sf->uiTop  = top;
+	sf->left = left;
+	sf->top  = top;
+
+	// Get all the audio data
+	unsigned char audio_tracks;
+	unsigned char audio_channels[7];
```

Comment:
The 7 occurs an awful lot in this code. Does it make sense to extract it into a meaningful constant?

Reply:
If smacker had a constant I would use it.
I decided to match the library.